File: tests/minifier/golden/python/closures_input.py
# Tests: closures, shadowing, nested functions, nonlocal.

def make_counter(start_value):
    current_count = start_value

    def increment(step_size):
        nonlocal current_count
        current_count = current_count + step_size
        return current_count

    def reset():
        nonlocal current_count
        current_count = start_value

    return increment, reset


def make_multiplier(factor_value):
    def multiply(input_number):
        # factor_value is a free variable from enclosing scope
        product_result = input_number * factor_value
        return product_result
    return multiply


def shadowing_test(outer_x):
    result_list = []
    for outer_x in range(outer_x):  # outer_x shadows the parameter here
        inner_val = outer_x * 2
        result_list.append(inner_val)
    return result_list


if __name__ == "__main__":
    inc, rst = make_counter(10)
    print(inc(1))   # 11
    print(inc(2))   # 13
    rst()
    print(inc(0))   # 10

    double = make_multiplier(2)
    triple = make_multiplier(3)
    print(double(5))   # 10
    print(triple(5))   # 15

    print(shadowing_test(4))  # [0, 2, 4, 6]

File: tests/minifier/golden/javascript/closures_input.js
// Tests: closures, nested scopes, var hoisting, let/const block scope.

function makeCounter(startValue) {
    let currentCount = startValue;

    function increment(stepSize) {
        currentCount = currentCount + stepSize;
        return currentCount;
    }

    function reset() {
        currentCount = startValue;
    }

    return { increment, reset };
}

function makeMultiplier(factorValue) {
    return function(inputNumber) {
        const productResult = inputNumber * factorValue;
        return productResult;
    };
}

const arrowMultiplier = (factorValue) => (inputNumber) => inputNumber * factorValue;

function varHoistingTest(limitValue) {
    // var is hoisted to function scope
    var collectedResults = [];
    for (var loopIndex = 0; loopIndex < limitValue; loopIndex++) {
        var doubledValue = loopIndex * 2;
        collectedResults.push(doubledValue);
    }
    return collectedResults;
}

const counter = makeCounter(10);
console.log(counter.increment(1));   // 11
console.log(counter.increment(2));   // 13
counter.reset();
console.log(counter.increment(0));   // 10

const double = makeMultiplier(2);
const triple = makeMultiplier(3);
console.log(double(5));   // 10
console.log(triple(5));   // 15

const arrowDouble = arrowMultiplier(2);
console.log(arrowDouble(7));  // 14

console.log(varHoistingTest(4));  // [0, 2, 4, 6]
